### ISSUE TITLE
Memory Leaks

### ISSUE THREAD

**Issue by MrPans** [dalu93/Each#4]
RunLoop retain Each's `_timer` property cause memory Leaks, Although `_timer` be marked  as `weak`.

```
RunLoop -> Timer -> Each
```

**Comment by dalu93** [comment]
@DeveloperPans I don't see any memory leaks in the code.
It's true, the run loop retains the timer which it is weak referenced by `Each`.

**Comment by MrPans** [comment]
the test step could expose memory leak are as follow:
1. We have two ViewController embed in `UINavigationController`: `ViewControllerA` and `ViewControllerB`
2. push to `ViewControllerB`
3. use `Each` like demo doing, at `ViewControllerB`'s `viewDidLoad` function.

```
    override func viewDidLoad() {
        super.viewDidLoad()
        Each(1).seconds.perform {
            print("second passed")
            return false
        }
    }
```
1. Pop `ViewControllerB`.
2. Timer is still running, because of the each instance wasn't free and `deInit` not been called.
3. We got a memory leak.

**Comment by dalu93** [comment]
For now, a simple workaround could be: store the `Each` instance in your `UIViewController` and call the `stop()` method in the `deinit` as below:

``` swift
class BViewController: UIViewController {

    let timer = Each(1).seconds

    deinit {
        timer.stop()
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        timer.perform { [weak self] in
            print("Timer called")
            self?.dismiss(animated: true, completion: nil)
            return .continue
        }
    }
}
```

**Comment by dalu93** [comment]
By the way, it works well even if you return `.stop/true` in the perform closure. This is quite normal, I don't see the need to return `.continue/false` when the closure deinitializes the owner (the view controller)

**Comment by MrPans** [comment]
Here is a way to resolve this issue in Objective-C: [PSWeakProxy](https://github.com/DeveloperPans/PSWeakProxy)
But invoke it case dependence.

I'm trying to find out a way with pure swift.

**Comment by dalu93** [comment]
@DeveloperPans I'm actually working on it on `feature/memory-leak-fix` branch

For now I succeeded to make a sort of weak timer. The only requirement is that the object has to hold a strong reference to the `Each` instance, like the example below:

``` swift
class BVC: UIViewController {

    let timer = Each(1).seconds

    override func viewDidLoad() {
        super.viewDidLoad()

        timer.perform {
            print("tiemer called")
            return .continue
        }
    }

    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)

        self.dismiss(animated: true, completion: nil)
    }
}
```

This works nicely and there is no need to call the `stop()` when the `BVC.deinit` is called

**Comment by MrPans** [comment]
great!

**Comment by dalu93** [comment]
So, as I explained in the post up here, with this fix, you are not able anymore to create a timer in your method scope by only typing the simple syntax (see below) because the controller has to keep a reference to it.

``` swift
Each(1).seconds.perform {
    // do something
    return .continue 
}
```

I found an another solution with a colleague of mine, which allows you to do it again. But `Each` needs the owner reference to check whether it's been deallocated in the mean time to stop the timer execution, so a prototype of the implementation will be

``` swift
Each(1).seconds.performOn(self) {
    // do something
    return .continue
}
```

What do you think guys? I can even try to wrap this method call up  in a `NSObject` extension.

**Comment by dalu93** [comment]
Actually I'm thinking about this issue really a lot. I decided that the memory leak should stay there: unfortunately it's the default behavior of `NSTimer` class provided by Apple. There are some workarounds to keep your code free from memory leaks using `Each`

## Workaround 1

In case you don't want to declare a property that holds the `Each` reference, create a normal `Each` timer in your method scope and return `.stop/true` whenever the `owner` instance is `nil`

``` swift
Each(1).seconds.perform { [weak self] in
    guard let _ = self else { return .stop }

    print("timer called")
    return .continue
}
```

90% of closures will call `self` somehow, so this isn't so bad

## Workaround 2

In case the first workaround wasn't enough, you can declare a property that holds the `Each` reference and call the `stop()` function whenever the `owner` is deallocated

``` swift
final class ViewController: UIViewController {
    private let _timer = Each(1).seconds

    deinit {
        _timer.stop()
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        _timer.perform {
            // do something and return. you can check here if the `self` instance is nil as for workaround #1
        }
    }
}
```

@DeveloperPans I'll update the `README` with these examples

**Comment by MrPans** [comment]
May be paste Workaround 2 to README could help more people and this discuss perform its values. Do you think so?

**Comment by dalu93** [comment]
I think for now it's ok until someone will find a good solution for it.
`Each` is cool because the developer can simply use it. All the fixes I found make the interface not easy as it is now.

**Comment by MrPans** [comment]
Less global variables are better… I think.

So I vote `performOn`: )
Best regards!

---

潘晟
Shawn Pan
Team Leader
http://shengpan.net

**Comment by dalu93** [comment]
I'm gonna implement that method too. In this way the developer can easily choose his implementation @DeveloperPans

**Comment by dalu93** [comment]
I just merged the new method on `develop`. It will be released soon in the version `1.1.1`
